rfcs/NNNN-preact-signals.md

```diff
@@ -0,0 +1,184 @@
+
+Preact signals exports the `effect()` method for this:
+
+```ts
+import {effect} from '@preact/signals-core';
+
+export const logSignal = (s: Signal<unknown>) =>
+  effect(() => {
+    // Run when someSignal changes
+    console.log(s.value);
+  });
+```
+
+We intend to offer three ways to automatically run signal access inside an effect:
+- A class mixin that runs the entire update lifecycle in an effect and causes the whole element to re-render upon changes.
+- A directive that applies a single signal to a binding
+- A customized `html` template tag that automatically applies the directive to signal-values objects.
+
+### SignalWatcher Mixin
+
+Conceptually, we want to run the reactive update lifecycle in an effect so that the signal library observes access to signals and trigger a new update.
+
+We can do this with an override of `performUpdate()` that wraps ReactiveElement's implementation in an effect:
+
+```ts
+    private _disposeEffect?: () => void;
+
+    override performUpdate() {
+      // ReactiveElement.performUpdate() also does this check, so we want to
+      // also bail early so we don't erroneously appear to not depend on any
+      // signals.
+      if (!this.isUpdatePending) {
+        return;
+      }
+      // If we have a previous effect, dispose it
+      this._disposeEffect?.();
+
+      // We create a new effect to capture all signal access within the
+      // performUpdate phase (update, render, updated, etc) of the element.
+      this._disposeEffect = effect(() => {
+        // When Signals change we need to re-render, but we need to get past
+        // the isUpdatePending in performUpdate(), so we set it to true.
+        this.isUpdatePending = true;
+        // We call super.performUpdate() so that we don't create a new effect
+        // only as the result of the effect running.
+        super.performUpdate();
+      });
+    }
+```
+
+An issue with this approach is that we bypass the reactive update lifecycle when signals change. We would like to integrate updates from signals and reactive properties into one batch. To do this we need separate watch and update code paths. Preact Signals don't have such an API, but we should be able to simulate it:
+
+```ts
+    override performUpdate() {
+      if (!this.isUpdatePending) {
+        return;
+      }
+      this._disposeEffect?.();
+      let updateFromLit = true;
+      this._disposeEffect = effect(() => {
+        if (updateFromLit) {
+          super.performUpdate();
+        } else {
+          // This branch is an effect run from Preact signals.
+          // This will cause another call into performUpdate, which will
+          // then create a new effect watching that update pass.
+          this.requestUpdate();
+        }
+        updateFromLit = false;
+      });
+    }
+```
+
+### watch() directive
```

Comment:
Let's include a section somewhere in the RFC about the impact to the analyzer's ability to typecheck values going through either `watch` or the `html` wrapper into templates, and how we might solve them